compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-amdaie-fuse-pack-into-for"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+class AMDAIEFusePackIntoForLoopPass
+    : public impl::AMDAIEFusePackIntoForLoopBase<
+          AMDAIEFusePackIntoForLoopPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<scf::SCFDialect>();
+  }
+
+  AMDAIEFusePackIntoForLoopPass() = default;
+  AMDAIEFusePackIntoForLoopPass(const AMDAIEFusePackIntoForLoopPass &pass){};
+  void runOnOperation() override;
+};
+
+void AMDAIEFusePackIntoForLoopPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  func::FuncOp funcOp = getOperation();
+  IRRewriter rewriter(context);
+
+  // Walk through the graph in post order and find the pack op.
+  TilingInterface packOp;
+  funcOp->walk<WalkOrder::PostOrder, ReverseIterator>([&](TilingInterface op) {
+    if (isa<linalg::FillOp>(op)) {
+      packOp = op;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+  if (!packOp) {
+    LLVM_DEBUG(llvm::dbgs() << "There is no pack op to be fused.\n");
+    return;
+  }
+
+  // Search the first use by a scf::ForOp user.
+  scf::ForOp forOp;
+  auto itProducerUses = llvm::find_if(packOp->getUses(), [&](OpOperand &use) {
+    forOp = dyn_cast<scf::ForOp>(use.getOwner());
+    return forOp;
+  });
+  if (!forOp || itProducerUses == packOp->getUses().end()) {
+    LLVM_DEBUG(llvm::dbgs() << "There is no for loop to fuse with.\n");
+    return;
+  }
+
+  // Search the producer slices accessed within the For op.
+  BlockArgument bbArg = forOp.getRegionIterArg(0);
+  SmallVector<tensor::ExtractSliceOp> sliceOps;
+  for (auto user : bbArg.getUsers()) {
+    if (auto genericOp = dyn_cast<linalg::GenericOp>(user)) {
+      for (auto [index, operand] : llvm::enumerate(genericOp.getOperands())) {
+        if (!isa<BlockArgument>(operand)) {
+          auto defOp = operand.getDefiningOp();
+          if (auto sliceOp = dyn_cast<tensor::ExtractSliceOp>(defOp)) {
+            sliceOps.push_back(sliceOp);
+          }
+        }
+      }
+    }
+  }
+
+  if (sliceOps.empty()) {
+    funcOp->emitOpError("There is no extract tensor slice.");
+    return signalPassFailure();
+  }
+
+  LoopLikeOpInterface loops = cast<LoopLikeOpInterface>(forOp.getOperation());
+
+  // Materialize each slice of the producer in place.
+  for (auto sliceOp : sliceOps) {
+    std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
+        scf::tileAndFuseProducerOfSlice(rewriter, sliceOp,
+                                        MutableArrayRef(&loops, 1));
+    if (!fusedProducer) {
+      funcOp->emitOpError("Failed to fuse pack ops into for loop.");
```

Comment:
This is misleading, the producer of slice is not necessarily a pack op, we are collecting all the sliceops in the loop and they can come from any kind of producer, and we are trying to fuse all of them here, not just pack ops

Reply:
Thanks for pointing this out. I modified the logic a little bit, but still constrains the producer as a tensor::PackOp to avoid undesired fusion...